### PR TITLE
JDK-8267690: Revisit (Doc)Tree search implemented by throwing an exception

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a moderately simple cleanup change, to eliminate using an exception to terminate scanning a `Tree` or `DocTree` in `(Doc)TreePath.getPath`.

The change is to set a field with the intended result, and once that field is set, do "best-effort" to eliminate any additional scanning.

It is easy enough to stop scanning items in a list, but it is not practical to totally stop scanning the subsequent sibling nodes, but we can substantially reduce the cost of scanning those nodes, and can definitely avoid scanning any children of those nodes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267690](https://bugs.openjdk.java.net/browse/JDK-8267690): Revisit (Doc)Tree search implemented by throwing an exception

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to 32aab7f0b116992b813a3759ebf44e8d8e3cbdb8
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8369/head:pull/8369` \
`$ git checkout pull/8369`

Update a local copy of the PR: \
`$ git checkout pull/8369` \
`$ git pull https://git.openjdk.java.net/jdk pull/8369/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8369`

View PR using the GUI difftool: \
`$ git pr show -t 8369`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8369.diff">https://git.openjdk.java.net/jdk/pull/8369.diff</a>

</details>

### CODE REVIEW

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Comment by pavelrappo** [comment]
You are not changing `(Doc)TreeScanner` API, but fixing this pair of internal `(Doc)TreePathScanner` subclasses; okay.

I think that your suggestion goes the extra mile as opposed to a more succinct suggestion by Jan [^1]. In particular, you avoid creating unnecessary `(Doc)TreePath` objects and calling `reduce`.

Could you have a look at other cases where we use `(Doc)TreeScanner` for search? For example:

  * com.sun.tools.javac.tree.TreeInfo#pathFor
  * com.sun.tools.javac.tree.TreeInfo#containsTypeAnnotation
  * jdk.jshell.ExpressionToTypeInfo#findExpressionPath
  
Maybe there's more. Is it possible to use your code there too? It would be good!

[^1]: https://bugs.openjdk.java.net/browse/JDK-8267690?focusedCommentId=14448893&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14448893

**Comment by pavelrappo** [comment]
I think I'm mistaken: your version does all these things I thought it wasn't doing. What's the difference between it and Jan's version? Is it more responsive wrt termination when the target node is found?

File: src/jdk.compiler/share/classes/com/sun/source/util/TreePath.java [hunk 1]
```diff
@@ -61,34 +61,44 @@ public static TreePath getPath(TreePath path, Tree target) {
...
-            @SuppressWarnings("serial") // Type of field is not Serializable
-            TreePath path;
-            Result(TreePath path) {
-                this.path = path;
+        class PathFinder extends TreePathScanner<TreePath,Tree> {
+            private TreePath result;
+
+
+            @Override
+            public TreePath scan(TreePath path, Tree docTree) {
```
Comment (pavelrappo) [inline src/jdk.compiler/share/classes/com/sun/source/util/TreePath.java:69]: The second parameter has nothing to do with `DocTree`. Copy-paste?

**Comment by jonathan-gibbons** [comment]
> I think I'm mistaken: your version does all these things I thought it wasn't doing. What's the difference between it and Jan's version? Is it more responsive wrt termination when the target node is found?

I didn't know/check Jan's version. If there's an equivalent version, I'm OK to share authorship credits, but not fussed either way.

File: src/jdk.compiler/share/classes/com/sun/source/util/TreePath.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/util/TreePath.java:69]: copy-paste error; will fix

**Review by pavelrappo — APPROVED** [review]

**Comment by jonathan-gibbons** [comment]
Thanks for the references to other candidates.

* com.sun.tools.javac.tree.TreeInfo#pathFor seems to be dead code

* com.sun.tools.javac.tree.TreeInfo#containsTypeAnnotation not a path finder, and looks like only called on small trees

* jdk.jshell.ExpressionToTypeInfo#findExpressionPath does seem to be another candidate for this work, but could/should be done separately

**Pull request closed without merging**